pkg/kubectl/role.go

```diff
@@ -0,0 +1,264 @@
+func (s *RoleGeneratorV1) StructuredGenerate() (runtime.Object, error) {
+	if err := s.parseAndValidateParams(); err != nil {
+		return nil, err
+	}
+
+	role := &rbac.Role{}
+	rule := rbac.PolicyRule{}
+
+	role.Name = s.Name
+	rule.Verbs = s.Verbs
+	rule.Resources = s.Resources
+	rule.APIGroups = s.APIGroups
+	rule.ResourceNames = s.ResourceNames
+
+	// At present, we only allow creating one single rule by using 'kubectl create role' command.
+	role.Rules = []rbac.PolicyRule{rule}
+
+	return role, nil
+}
+
+// parseAndValidateParams validates required fields are set to support structured generation.
+func (s *RoleGeneratorV1) parseAndValidateParams() error {
+	if len(s.Name) == 0 {
+		return fmt.Errorf("name must be specified")
+	}
+
+	err := s.parseAndValidateVerbs()
+	if err != nil {
+		return err
+	}
+
+	err = s.parseAndValidateResources()
+	if err != nil {
+		return err
+	}
+
+	return s.parseAndValidateResourceNames()
+}
+
+// parseAndValidateVerbs parses and validates verbs.
+func (s *RoleGeneratorV1) parseAndValidateVerbs() error {
+	// support specify multiple verbs together
+	// e.g. --verb=get,watch,list
+	verbs := []string{}
+	for _, v := range s.Verbs {
+		verbs = mergeArrays(verbs, strings.Split(v, ","))
+	}
+
+	if len(verbs) == 0 {
+		return fmt.Errorf("at least one verb must be specified")
+	}
+
+	containsVerbAll := false
+	for _, v := range verbs {
+		if !arrayContains(validResourceVerbs, v) && !arrayContains(validNonResourceVerbs, v) {
+			return fmt.Errorf("invalid verb: '%s'", v)
+		}
+		if v == "*" {
+			containsVerbAll = true
+		}
+	}
+
+	// VerbAll respresents all kinds of verbs.
+	if containsVerbAll {
+		s.Verbs = []string{"*"}
+	} else {
+		s.Verbs = verbs
+	}
+	return nil
+}
+
+// parseAndValidateResources parses and validates resources.
+func (s *RoleGeneratorV1) parseAndValidateResources() error {
+	// support specify multiple resources together
+	// e.g. --resource=pods,deployments
+	candidateResources := []string{}
+	for _, r := range s.Resources {
+		candidateResources = mergeArrays(candidateResources, strings.Split(r, ","))
+	}
+
+	if len(candidateResources) == 0 {
+		return fmt.Errorf("at least one resource must be specified")
+	}
+
+	if s.Mapper == nil {
+		return fmt.Errorf("no REST Mapper found to validate resources")
+	}
+
+	resources := []string{}
+	for _, r := range candidateResources {
+		// support resource.group pattern
+		index := strings.Index(r, ".")
+		resourceName, apiGroup := "", ""
+
+		// No API Group specified, use "" as core API Group
+		if index == -1 {
+			resourceName = r
+		} else {
+			resourceName, apiGroup = r[0:index], r[index+1:]
+		}
+
+		resource, err := s.Mapper.ResourceFor(schema.GroupVersionResource{Resource: resourceName})
+		if err != nil {
+			return err
+		}
+		if resource.Resource != resourceName {
+			// Transfer from short form name to long form name.
+			longResourceName, _ := ResourceLongFormFor(resourceName)
```

Comment:
Yes, generators should not deal with restmapping, especially since we treat them as api and intend to move them server-side.